include/swift/AST/DiagnosticsSema.def

```diff
@@ -394,6 +394,51 @@ ERROR(cannot_convert_closure_result_protocol,none,
 ERROR(cannot_convert_closure_result_nil,none,
       "'nil' is not compatible with closure result type %0", (Type))
 
+ERROR(cannot_convert_generic_type_argument,none,
+      "cannot convert %0 to expected argument type %1, "
+      "arguments to generic parameter %2 (%3 and %4) are expected to be equal",
```

Comment:
Overall after looking at all of the diagnostic changes it seems like it might be test to have original messages with added notes about mismatched generic parameters, WDYT?

Reply:
I think that probably is the best approach. My one question is would there be too many notes for this diagnostic to be useful? We'd be emitting one note pointing to the mismatch at the location itself, and one note pointing to the declaration of the generic parameter per mismatch which i feel could get confusing.

Reply:
I think we can remove `declared here` note in favor of just pointing new note to the parameter declaration then? So in example from description it would look like:

```
example.swift:4:30: error: cannot assign value of type 'A<Int, Int, Int>' to type 'A<Bool, Int, Float>'
let b: A<Bool, Int, Float> = a
                             ^
example.swift:1:10: note: arguments to generic parameter 'X' ('Int' and 'Bool') are expected to be equal
struct A<X, Y, Z> {}
         ^
```

or to the type repr location e.g.

```
example.swift:4:30: error: cannot assign value of type 'A<Int, Int, Int>' to type 'A<Bool, Int, Float>'
let b: A<Bool, Int, Float> = a
                             ^
example.swift:1:10: note: arguments to generic parameter 'X' ('Int' and 'Bool') are expected to be equal
let b: A<Bool, Int, Float> = a
         ^
```

Reply:
So is the idea that we would put the diagnostic at the first one where possible, and the second one if that source location wasn't valid? If so, I think that seems good!

Reply:
Sure, that sounds good to me.